sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+    public Write withPartitionKey(String partitionKey) {
+      return builder().setPartitionKey(partitionKey).build();
+    }
+
+    /**
+     * Allows to specify custom implementation of {@link KinesisPartitioner}.
+     *
+     * <p>This method should be used to balance a distribution of new written records among all
+     * stream shards.
+     *
+     * <p>Using one of the methods {@link KinesisIO.Write#withPartitioner(KinesisPartitioner)} or
+     * {@link KinesisIO.Write#withPartitionKey(String)} is required but not both in the same time.
+     */
+    public Write withPartitioner(KinesisPartitioner partitioner) {
+      return builder().setPartitioner(partitioner).build();
+    }
+
+    /**
+     * Specify the configuration properties for Kinesis Producer Library (KPL).
+     *
+     * <p>Example of creating new KPL configuration:
+     *
+     * {@code
+     * Properties properties = new Properties();
+     * properties.setProperty("CollectionMaxCount", "1000");
+     * properties.setProperty("ConnectTimeout", "10000");}
+     */
+    public Write withProducerProperties(Properties properties) {
+      return builder().setProducerProperties(properties).build();
+    }
+
+    /**
+     * Allows to specify custom {@link AWSClientsProvider}. {@link AWSClientsProvider} creates new
+     * {@link IKinesisProducer} which is later used for writing to Kinesis.
+     *
+     * <p>This method should be used if
+     * {@link Write#withAWSClientsProvider(String, String, Regions)} does not suit well.
+     */
+    public Write withAWSClientsProvider(AWSClientsProvider awsClientsProvider) {
+      return builder().setAWSClientsProvider(awsClientsProvider).build();
+    }
+
+    /**
+     * Specify credential details and region to be used to write to Kinesis. If you need more
+     * sophisticated credential protocol, then you should look at {@link
+     * Write#withAWSClientsProvider(AWSClientsProvider)}.
+     */
+    public Write withAWSClientsProvider(String awsAccessKey, String awsSecretKey, Regions region) {
+      return withAWSClientsProvider(awsAccessKey, awsSecretKey, region, null);
+    }
+
+    /**
+     * Specify credential details and region to be used to write to Kinesis. If you need more
+     * sophisticated credential protocol, then you should look at {@link
+     * Write#withAWSClientsProvider(AWSClientsProvider)}.
+     *
+     * <p>The {@code serviceEndpoint} sets an alternative service host. This is useful to execute
+     * the tests with Kinesis service emulator.
+     */
+    public Write withAWSClientsProvider(
+        String awsAccessKey, String awsSecretKey, Regions region, String serviceEndpoint) {
+      return withAWSClientsProvider(
+          new BasicKinesisProvider(awsAccessKey, awsSecretKey, region, serviceEndpoint));
+    }
+
+    /**
+     * Specify the number of retries that will be used to flush the outstanding records in
+     * case if they were not flushed from the first time. Default number of retries is
```

Comment:
I'd recommend to not allow a user to configure this, for all the usual reasons why other transforms in Beam prefer not to expose tuning knobs. The default value is likely good enough for all cases.

Reply:
Well, I agree that if it's a default rule for Beam transforms then it makes sense to drop it. But from the user point of view, I'd prefer to have some knobs in case when something goes not as expected.

Reply:
I understand that in general this is up to debate, but in Beam the rule of thumb is to expose a tuning knob only as a last resort: if someone could conceivably end up in a situation where their pipeline exhibits catastrophically poor behavior with the default value AND if it seems infeasible to choose a proper value automatically in general. I don't think this applies here - so for consistency with the rest of Beam codebase, I'd like to ask to remove this knob.

For context, the justification for this rule of thumb is that we've found that, whenever there is a tuning knob, nearly all users set it to a bad value - or perhaps to a value that used to be good at the moment they chose it, but not anymore (e.g. because the pipeline changed, the data changed, implementation of underlying libraries changed, the runner changed, etc) - and the user doesn't have a process in place to keep regularly manually re-tuning the parameter for every use case (I've never seen a user who has such a process).

E.g. inside Google, the legacy MapReduce used to have hundreds of tuning parameters, and when a user complained about a performance problem, often the first recommendation was "have you tried removing all tuning parameters" and it very often helped a great deal :) When a value is explicitly specified, we have to respect it and can't automatically make things better.

Reply:
Thank you for so detailed explanation, it makes sense for me.  I agree to not provide this knob to end user but I'd keep it for using in unit test - I want to test throwing an exception when flush is failing and I want to reduce a number of retries to minimum, otherwise this UT takes long time to finish. 
So, I changed the visibility of `withRetries()` to package instead of public (like `withBigtableService()` in `BigtableIO`). 
Are you ok with this?